async/runtime/src/openxla/runtime/async/async_runtime_cc.h

```diff
@@ -0,0 +1,104 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef OPENXLA_RUNTIME_ASYNC_ASYNCRUNTIME_CC_H_
+#define OPENXLA_RUNTIME_ASYNC_ASYNCRUNTIME_CC_H_
+
+#include "iree/base/status_cc.h"
+#include "iree/vm/ref_cc.h"
+#include "tensorflow/tsl/concurrency/async_value.h"
+#include "tensorflow/tsl/concurrency/async_value_ref.h"
+#include "tensorflow/tsl/concurrency/chain.h"
+
+struct iree_async_value_t;
+
+namespace openxla::runtime::async {
+
+// Returns the  IREE VM value type (eg, F32) corresponding to the given
+// template parameter native type (eg, float).
+template <typename NativeT>
+iree_vm_value_type_t NativeToVMValueType() {
+  // Make the expression depend on the template parameter NativeT so
+  // that this compile-time error only appears if this function is
+  // instantiated with some concrete type that is not specialized
+  // below.
+  static_assert(!std::is_same<NativeT, NativeT>::value,
+                "Cannot map native type to vm type.");
+  return IREE_VM_VALUE_TYPE_NONE;
+}
+
+template <>
+inline iree_vm_value_type_t NativeToVMValueType<int8_t>() {
+  return IREE_VM_VALUE_TYPE_I8;
+}
+
+template <>
+inline iree_vm_value_type_t NativeToVMValueType<int16_t>() {
+  return IREE_VM_VALUE_TYPE_I16;
+}
+
+template <>
+inline iree_vm_value_type_t NativeToVMValueType<int32_t>() {
+  return IREE_VM_VALUE_TYPE_I32;
+}
+
+template <>
+inline iree_vm_value_type_t NativeToVMValueType<int64_t>() {
+  return IREE_VM_VALUE_TYPE_I64;
+}
+
+template <>
+inline iree_vm_value_type_t NativeToVMValueType<float>() {
+  return IREE_VM_VALUE_TYPE_F32;
+}
+
+template <>
+inline iree_vm_value_type_t NativeToVMValueType<double>() {
+  return IREE_VM_VALUE_TYPE_F64;
+}
+
+template <typename T>
+using EnableIfScalarType = typename std::enable_if_t<
+    std::disjunction_v<std::is_same<T, float>, std::is_same<T, double>,
+                       std::is_same<T, int8_t>, std::is_same<T, int16_t>,
+                       std::is_same<T, int32_t>, std::is_same<T, int64_t>>>;
+
+class AsyncValue : public iree::vm::RefObject<AsyncValue> {
```

Comment:
```
struct iree_async_value_t {
  iree_hal_resource_t resource;
  iree_vm_value_type_t type_;
  tsl::RCReference<tsl::AsyncValue> value_;
}
```

I think this is the C-way of doing things in IREE, but it looks a little bit hal specific, did you look into it? Was it a deliberate decision to go with vm::RefObject? Any pros and cons of different approaches?

Given that everything is mostly IREE-C-style, I'd look into this option, it should be roughly the same at run time.

You also can make it pure C with forward declaration and a little bit of manual ref counting.

```

void drop_tsl_ref(void* tsl_ref);

struct iree_async_value_t {
  ...
  void* value_; // or maybe add a C typedef
};
```

And in `iree_async_value_t` destructor you'll just have to explicitly call `value->DropRef()` instead of relying on RCReference. Or as in the example implement `drop_tsl_ref` in `async_runtime.cpp`. And fancy C++ move constructors will be moved to C++ implementation as helper functions.